openharmony: add instructions for servoshell (#24)
servo/book: src/hacking/building-for-openharmony.md

Building for OpenHarmony requires the following:

1. The OpenHarmony SDK. This is sufficient to compile servo as a shared library for OpenHarmony.

> 2. The `hvigor` build tool to compile apps into an app bundle and sign it.

Reviewer: ```suggestion
2. The `hvigor` build tool to compile an application into an app bundle and sign iy.
```

`hvigor` (not the wrapper `hvigorw`) is also available via `npm`.
1. Install the same nodejs version as the commandline-tools ship.

>    For HarmonyOS NEXT Node 18 is shipped. Ensure node binary is in PATH.

Reviewer: ```suggestion
   For HarmonyOS NEXT Node 18 is shipped. Ensure that the `node` binary is in PATH.
```

2. Edit your `.npmrc` to contain the following line:

    ```
    @ohos:registry=https://repo.harmonyos.com/npm/
    ```

3. Install hvigor and the hvigor-ohos-plugin. This will create a `node_modules` directory in the current directory.
   ```commandline
   npm install @ohos/hvigor
   npm install @ohos/hvigor-ohos-plugin
   ```
4. Now you should be able to run `hvigor.js` in your OpenHarmony project to build a hap bundle:
   ```
   /path/to/node_modules/@ohos/hvigor/bin/hvigor.js assembleHap
   ```

### Configuring hdc on Linux

`hdc` is the equivalent to `adb` for OpenHarmony devices.
You can find it in the `toolchains` directory of your SDK.
For convenience purposes, you might want to add `toolchains` to your `PATH`.

> Among others,`hdc` can be used to open a shell or send/receive files from a device.

Reviewer: ```suggestion
Among others,`hdc` can be used to open a shell or transfer files between a device and the host system.
```

> Among others,`hdc` can be used to open a shell or transfer files between a device and the host system.

Reviewer 2: ```suggestion
Among others, `hdc` can be used to open a shell or transfer files between a device and the host system.
```

1. Tap the build number multiple times to enable developer mode.
2. Then navigate to the developer options and enable USB debugging.
3. When you connect your device for the first time, confirm the pop-up asking you if you want to trust the computer you are connecting to.

## Signing configuration

Most devices require that the HAP is digitally signed by the developer to be able to install it.
When using the `hvigor` tool, this can be accomplished by setting a static `signingConfigs` object in the `build-profile.json5` file or by dynamically creating the `signingConfigs` array on the application context object in the `hvigorfile.ts` script.

The `signingConfigs` property is an array of objects with the following structure:

```json
{
    "name": "default",
    "type": "<OpenHarmony or HarmonyOS>",
    "material": {
        "certpath": "/path/to/app-signing-certificate.cer",
        "storePassword": "<encrypted password>",
        "keyAlias": "debugKey",
        "keyPassword": "<encrypted password>",
        "profile": "/path/to/signed-profile-certificate.p7b",
        "signAlg": "SHA256withECDSA",
        "storeFile": "/path/to/java-keystore-file.p12"
    }
}
```

> Here <encrypted password> is a hexadecimal string representation of the plaintext password after being encrypted.

Reviewer 2: ```suggestion
Here `<encrypted password>` is a hexadecimal string representation of the plaintext password after being encrypted.
```

To check if `hdc` is now working, you can run `hdc list targets` and it should show your device serial number.
If it doesn't work, try rebooting.

Please note, that your device needs to be in "Developer mode" with USB debugging enabled.
The process here is exactly the same as one android:
1. Tap the build number multiple times to enable developer mode.
2. Then navigate to the developer options and enable USB debugging.
3. When you connect your device for the first time, confirm the pop-up asking you if you want to trust the computer you are connecting to.

## Signing configuration

Most devices require that the HAP is digitally signed by the developer to be able to install it.
When using the `hvigor` tool, this can be accomplished by setting a static `signingConfigs` object in the `build-profile.json5` file or
by dynamically creating the `signingConfigs` array on the appication context object using the `hvigorfile.ts` script.

The `signingConfigs` property is an array of objects with the following structure:

```json

{
    "name": "default",

>     "type": "OpenHarmony",

Author: I've changed this to `"type": "<OpenHarmony or HarmonyOS>"`

    "material": {
        "certpath": "/path/to/app-signing-certificate.cer",
        "storePassword": "<encrypted password>",
        "keyAlias": "debugKey",
        "keyPassword": "<encrypted password>",
        "profile": "/path/to/signed-profile-certificate.p7b",
        "signAlg": "SHA256withECDSA",
        "storeFile": "/path/to/java-keystore-file.p12"
    }
}
```

Here <encrypted password> is a hexadecimal string representation of the plaintext password after being encrypted.
The key and salt information used to encrypt these passwords are generated by DevEco Studio IDE and are stored on-disk alongside
the certificate files and keystore, usually under `<USER HOME>/.ohos/config/openharmony`.

> You can use the IDE to generate information needed for password encryption, the required application and

Reviewer 3: Probably we should mention that this is only for developing / testing purposes. To ship an app, one would need to get the signature configuration from the app store provider.

Author: done in [4090da6](https://github.com/servo/book/pull/24/commits/4090da6a8f2698c7c1b68402252580a44c5187bd)

profile certificate files, and the keystore itself using:

1. Open Project Structure dialog from `File > Project Structure` menu.
2. Under the 'Signing Config' tab, enable the 'Automatically generate signature'.

Once generated, it is necessary to point `mach` to the above "signing material" configuration using the
`SERVO_OHOS_SIGNING_CONFIG` environment variable. The value of the variable must be a file path to a
valid `.json` file with the same structure as the `signingConfigs` property given above, but with `certPath`, `storeFile` and
`profile` given as *paths relative to the json file*, instead of absolute paths.

## Building Servo shell

Before building servo, ensure the following environment variables are set in the current shell:

```commandline

>     # Note: The openharmony sdk is under ${DEVECO_SDK_HOME}/HarmonyOS-NEXT-${HOS_VERSION}/openharmony

Reviewer 3: I think in Developer Beta 7 / Beta 1 this path was changed. I'd need to check next week, but I think it's now something like `${DEVECO_SDK_HOME}/default/openharmony`.

Author: I've update the text to be `${DEVECO_SDK_HOME}/default/openharmony`. Is that different from one where commandline tools is used?

Reviewer 3: `DEVECO_SDK_HOME` points to `/path/to/DevEcoStudio/sdk`, so that seems to match.

    # Presumably you would need to replicate this directory structure
    export OHOS_SDK_NATIVE=/path/to/openharmony-sdk/platform/api-version/native
    export OHOS_BASE_SDK_HOME=/path/to/openharmony-sdk/platform
    export HVIGOR_PATH=/path/to/parent/directory/containing/node_modules # not required if hvigorw is in PATH
    export SERVO_OHOS_SIGNING_CONFIG=/path/to/signing-configs.json # required if the HAP must be signed.
```

The following command can then be used to compile the servo shell application for a 64-bit ARM device:

```commandline
./mach build --ohos --release
```

In `mach build`, `mach install` and `mach package` commands, `--ohos` is an alias for `--target aarch64-unknown-linux-ohos`.

> To build for the emulator, use `--target x86_64-unknown-linux-ohos`

Author: Good catch! I've made the text more explicit that this command is for an emulator running on x86-64 hosts.
